### ISSUE TITLE
Icons and Popups on headers.

### ISSUE THREAD

**Issue by ddude** [ocornut/imgui#316]
Hello, I'd like to add more icons to the right of a header line but don't seem to find how.

Also I'd like to show a popup menu using ```ImGui::BeginPopupContextItem``` when the header line is right-clicked but it doesn't work.

Thanks! Loving the library so far :)

**Comment by ocornut** [comment]
Jeremie, what do you mean by header line?

**Comment by ddude** [comment]
Ah sorry I didn't specify, I meant the title bar of windows from calling ```ImGui::Begin``` as well as the one from ```ImGui::CollapsingHeader```.

**Comment by ocornut** [comment]
You can already use BeginPopupContextItem() or any `*Item*` call on a Collapsing Header.

You can't directly do that for the title bar but it's easy to replicate the content of BeginPopupContentItem() and adjust it for a custom rectangle.

```
#include <imgui_internal.h> // for GetCurrentWindow()->TitleBarRect9)

if (ImGui::IsWindowHovered() && ImGui::IsMouseClicked(1))
{
    ImRect rect = ImGui::GetCurrentWindow()->TitleBarRect();
    if (ImGui::IsMouseHoveringRect(rect.Min, rect.Max, false))
        ImGui::OpenPopup("FOO");
}
```

You'll need to update as I added a third parameter to IsMouseHoveredRect() to ignore the current clipping setting. Otherwise you can do

**Comment by ocornut** [comment]
For icons you can't really trivially, but you can use various approaches.

a) Include icons in your font and use them in your title, see https://github.com/ocornut/imgui/issues/232
b) Draw them manually after the fact. Maybe add spacing in the title then use GetWindowDrawList()->AddImage(). For the title bar because the clipping region doesn't include the title bar you'd need to, e.g. PushClipRectFullScreen(), PopClipRect() and draw the icons in-between.
c) Copy and recreate custom functions based on existing CollapsingHeader()

If you need the icons to be interactive then maybe it is a little more work.

**Comment by ddude** [comment]
Thanks for the info!

The context menu does indeed work on collapsing headers, turns out I only tested on a window and assumed headers worked the same way;)

I do indeed want icons to act like buttons; basically I'll be using this for flags I still want to toggle when the window or section is collapsed.

I'll get back to you over the weekend for the rest, I'm actually using your library from D. I'll need to fork and update both CIMGUI and derelict-imgui before I can try this code ;)

Just as a quick idea, how about splitting both windows and sections into separate begin/end calls for both their header and content, something like this:

    ImGui::BeginHeader(...);
    // either draw complete header or append to arrow/label depending on args to previous call.
    if (ImGui::Begin(...)) {
        // draw window contents
    }
    ImGui::End();

Curious to know what you think!

**Comment by ocornut** [comment]
We can't break the existing API but manipulating the header is basically as simple as changing the clipping rectangle, moving the cursor, doing what you need then restoring the cursor. So that possibly could be turned into a high-level helper but before that we'd need to try low level.

This app has been doing their own title bar https://github.com/ApoorvaJ/Papaya

**Comment by ddude** [comment]
I can't seem to get the new version of IMGUI to work. The shared library loads fine but nothing appears on the screen. If I downgrade to v1.44 everything works again.

**Comment by ocornut** [comment]
Please open a new thread. I don't know what "shared library" you are talking about, if it is a concept of cimgui or dimgui wrappers.

You can git bisect to try to find a bug, but frankly I don't think any of the changes since 1.44 would have broken something of this nature, so I'd guess maybe some of those wrappers weren't rebuild.

**Comment by ddude** [comment]
Just found it, turns out e58f99179aa3e8fa78e783f772c31ee23d40863d added the ```DisplayFramebufferScale``` field and the equivalent declaration in D wasn't updated.

I'm almost to the point where I can try these samples in D, I'll let you know how it goes.

**Comment by ocornut** [comment]
Oh I didn't realize the D library had to replicate every data structure.

**Comment by ddude** [comment]
Yeah, it is ABI compatible with C but not source compatible. Its most likely that way so there's zero runtime overhead yet the language isn't burdened with legacy C syntax.

**Comment by ddude** [comment]
Alright I've got the code for window context menus to work, except no popup is showing. I traced the execution and ```OpenPopup``` is properly called when I right-click the title bar, ```BeginPopup``` returns true but no menu on screen. I tried ```BeginPopupContextItem``` with the same results.

Is there something I'm missing?

**Comment by ocornut** [comment]
Hard to say without detailed information. What's your code like?

**Comment by ddude** [comment]
Damn I feel stupid hehe; I found my problem while pasting the code here and stripping it of application logic. Turns out I was calling ```BeginPopup``` only on the frame the right-click happened. It works great now!

I'll try the icons now and see if I can make them clickable with tooltips.

**Comment by ocornut** [comment]
Hi Jeremie,

I have recently added missing high-level version of ImGui::PushClipRect / ImGui::PopClipRect in imgui_internal.h which may be helpful here.
https://github.com/ocornut/imgui/commit/40ddf8e2870d397a2673b12c05519a0487f48501

The difference with draw_list->PushClipRect() is that the later would only affect rendering while the earlier also affect how hit tests and widget culling are performed.

Have you moved forward with this idea? Let me know if this issue can be closed or if there are other desirable changes.

**Comment by ocornut** [comment]
Adding this back to my mental list today as I'm looking at tree and collapsing headers. Will see how we may be able to make it easier to add custom icons to a collapsing headers. (See #600)

**Comment by azonenberg** [comment]
> You can already use BeginPopupContextItem() or any `*Item*` call on a Collapsing Header.
> 
> You can't directly do that for the title bar but it's easy to replicate the content of BeginPopupContentItem() and adjust it for a custom rectangle.
> 
> ```
> #include <imgui_internal.h> // for GetCurrentWindow()->TitleBarRect9)
> 
> if (ImGui::IsWindowHovered() && ImGui::IsMouseClicked(1))
> {
>     ImRect rect = ImGui::GetCurrentWindow()->TitleBarRect();
>     if (ImGui::IsMouseHoveringRect(rect.Min, rect.Max, false))
>         ImGui::OpenPopup("FOO");
> }
> ```
> 
> You'll need to update as I added a third parameter to IsMouseHoveredRect() to ignore the current clipping setting. Otherwise you can do

This example code works fine if the window is not docked, but fails if the window is docked. Is there a way to handle a right click event on a docked window's tab?
